---
description: >-
  Fiber supporte la gestion des erreurs centralis√©e en passant un argument d'erreur dans la m√©thode Next qui vous permet de consigner des erreurs √† des services externes ou d'envoyer une r√©ponse HTTP personnalis√©e au client.
---

# üêõ Gestion des erreurs

## Erreurs de capture

Il est important de s‚Äôassurer que la Fibre d√©tecte toutes les erreurs qui se produisent lors de l‚Äôex√©cution des gestionnaires de routes et des middleware. Vous devez les passer au ctx `.La fonction Next()` , o√π les fibres les attraperont et les traiteront.

{% tabs %}
{% tab title="Example" %}
```go
applicationGet("/", func(c *fiber.Ctx) {
    err := c.SendFile("file-does-not-exist")

    if err != nil {
        c.Next(err) // Passer l'erreur √† la Fibre
    }
})
```
{% endtab %}
{% endtabs %}

Fibre ne g√®re pas les paniques [](https://blog.golang.org/defer-panic-and-recover) par d√©faut. Pour r√©cup√©rer d'une panique √©mise par n'importe quel gestionnaire dans la pile, vous devez inclure le middleware `Recover` comme montr√© ci-dessous:

{% code title="Example" %}
```go
import du paquet principal

(
    "github.com/gofiber/fiber"
    "github.com/gofiber/fiber/middleware"
)

func main() {
    app := fiber.Application

    New() .Utiliser(middleware.R√©cup√©rer ())

    l'application.Get("/", func(c *fiber.Ctx) {
        panique ("Cette panique est attrap√©e par le ErrorHandler")
    })

    journal.Fatal(app.√âcoute(3000))
}

```
{% endcode %}

Parce que `ctx.Next()` accepte une interface `erreur` , vous pouvez utiliser la structure d'erreur personnalis√©e de Fiber pour passer un `statuscode` suppl√©mentaire en utilisant `fibre.NewError()`. Il est facultatif de passer un message si ce champ est laiss√© vide, le message par d√©faut \(`404` √©quivaut √† `Non Trouv√©`\).

{% code title="Example" %}
```go
applicationGet("/", func(c *fiber.Ctx) {
    err := fibre.NewError(503)
    c.Next(err) // 503 Service Indisponible

    err := fibre.NewError(404, "D√©sol√©, pas trouv√© !")
    c.Next(err) // 404 D√©sol√©, pas trouv√©!
})
```
{% endcode %}

## Gestionnaire d'erreur par d√©faut

Fiber fournit un gestionnaire d'erreur par d√©faut. Pour une erreur standard, la r√©ponse est envoy√©e en tant que **500 Erreur interne du serveur**. Si l'erreur est de type [fiber\*Error](https://godoc.org/github.com/gofiber/fiber#Error), la r√©ponse est envoy√©e avec le code de statut et le message fourni.

{% code title="Example" %}
```go
// Ceci est l'application
de gestion des erreurs par d√©faut.Param√®tres.Gestionnaire d'erreurs = func(ctx *Ctx, err error) {
    // Statuscode par d√©faut √† 500
    code := StatusInternalServerError

    // Retire le code d'√©tat personnalis√© s'il s'agit d'une fibre. Erreur
    si e, ok := err.(*Erreur); ok {
        code = e.Code
    }

    // Retourne une r√©ponse HTTP
    ctx.Statut(code).SendString(err.Erreur())
}
```
{% endcode %}

## Gestionnaire d'erreurs personnalis√©

Le gestionnaire d'erreurs personnalis√© peut √™tre d√©fini via l'application `.Param√®tres.ErrorHandler`

Dans la plupart des cas, le gestionnaire d'erreur par d√©faut devrait √™tre suffisant. Cependant, un gestionnaire d'erreurs personnalis√© peut vous √™tre utile si vous voulez capturer diff√©rents types d'erreurs et prendre des mesures en cons√©quence. . Envoyez un courriel de notification ou une erreur de log √† un syst√®me centralis√©. Vous pouvez √©galement envoyer une r√©ponse personnalis√©e au client, par exemple une page d'erreur ou juste une r√©ponse JSON.

L'exemple suivant montre comment afficher les pages d'erreur pour diff√©rents types d'erreurs.

{% code title="Example" %}
```go
app := fibre.New()

// D√©finition d'un gestionnaire d'erreurs personnalis√©
application.Param√®tres.Gestionnaire d'erreurs = func(ctx *Ctx, err error) {
    // Statuscode par d√©faut √† 500
    code := StatusInternalServerError

    // Retire le code d'√©tat personnalis√© s'il s'agit d'une fibre. Erreur
    si e, ok := err.(*Erreur); ok {
        code = e.Code
    }

    // Envoyer une page d'erreur personnalis√©e
    erreur := ctx.Envoyer un fichier (fmt.Sprintf("./%d.html", code))
    si errr != nil {
        ctx.Statut(500).SendString("Erreur de serveur interne")
    } autre {
        ctx.Statut(code)
    }
}
```
{% endcode %}

> Remerciements sp√©ciaux au framework [Echo](https://echo.labstack.com/) & [Express](https://expressjs.com/) pour l'inspiration concernant la gestion des erreurs.

